### PR TITLE
Angular 11 and Linter Upgrades

### PR DESCRIPTION
**Pull request by adandach**
## Overview
Upgrade from Angular 9 to Angular 11
Replace tslint (deprecated) with eslint
Replace sass-lint (deprecated and unmaintained) with stylelint
Introduces prettier for formatting
Create a pre-commit hook to run formatting / linting against staged files before they are committed

#### Angular upgrade
- Run `yarn unit`

#### Formatter / Linter  - Changes that can automatically be fixed
- Modify a file to violate the linter/formatter rules. (add an extra line break in a scss file)
- Try to commit the file
- The file will be updated to match the linter / formatter rules prior to being committed

#### Formatter / Linter  - Changes that cannot automatically be fixed
- Modify a file to violate the linter/formatter rules. (change the selector prefix from `rq` to `app` in `rq-page.component.ts` for example)
- Try to commit the file
- The commit will fail with an error

## Note
All spec files were formatted when the Jest PR was merged into this branch

### CODE REVIEW

**Comment by rkennel** [comment]
There are a lot more warnings now than there used to be when running `yarn install`.  Should we address these warnings or separate that into another issue? 

**Current Main Branch**
```
rkennel@mainframe:~/Workspaces/prs/compare/retroquest/ui$ yarn install
yarn install v1.22.5
[1/4] Resolving packages...
warning jest > jest-cli > jest-config > jest-environment-jsdom > jsdom > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning jest > jest-cli > jest-config > jest-environment-jsdom > jsdom > request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
[2/4] Fetching packages...
info fsevents@1.2.13: The platform "linux" is incompatible with this module.
info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.1.3: The platform "linux" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.3.2: The platform "linux" is incompatible with this module.
info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > ng-recaptcha@5.0.0" has incorrect peer dependency "@angular/core@^6.0.0 || ^7.0.0 || ^8.0.0".
warning " > ts-jest@26.5.4" has incorrect peer dependency "typescript@>=3.8 <5.0".
[4/4] Building fresh packages...
success Saved lockfile.
Done in 16.77s.
```

**Pull Request #280**
```
rkennel@mainframe:~/Workspaces/prs/retroquest/ui$ yarn install
yarn install v1.22.5
info No lockfile found.
[1/4] Resolving packages...
warning @angular/http@6.1.10: Package no longer supported. Use @angular/common instead, see https://angular.io/guide/deprecations#angularhttp
warning core-js@2.6.12: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.
warning npm > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning npm > node-gyp > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning npm > request > har-validator@5.1.5: this library is no longer supported
warning @angular-devkit/build-angular > html-webpack-plugin@3.2.0: 3.x is no longer supported
warning @angular-devkit/build-angular > istanbul@0.4.5: This module is no longer maintained, try this instead:
  npm i nyc
Visit https://istanbul.js.org/integrations for other alternatives.
warning @angular-devkit/build-angular > node-sass > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning @angular-devkit/build-angular > node-sass > node-gyp > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning @angular-devkit/build-angular > @angular-devkit/core > chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
warning @angular-devkit/build-angular > webpack-dev-server > chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
warning @angular-devkit/build-angular > uglifyjs-webpack-plugin > uglify-es@3.3.9: support for ECMAScript is superseded by `uglify-js` as of v3.13.0
warning @angular-devkit/build-angular > @angular-devkit/core > chokidar > fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
warning @angular-devkit/build-angular > webpack > watchpack > watchpack-chokidar2 > chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
warning @angular-devkit/build-angular > istanbul-instrumenter-loader > istanbul-lib-instrument > babel-generator > babel-runtime > core-js@2.6.12: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.
warning @angular-devkit/build-angular > stylus > css-parse > css > urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
warning @angular-devkit/build-angular > stylus > css-parse > css > source-map-resolve > urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
warning @angular-devkit/build-angular > stylus > css-parse > css > source-map-resolve > resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
warning @angular/cli > @schematics/update > npm-registry-client > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning @angular/compiler-cli > chokidar@1.7.0: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
warning @angular/compiler-cli > chokidar > fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
warning karma > chokidar@1.7.0: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
warning karma > core-js@2.6.12: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.
warning karma > socket.io > socket.io-parser > json3@3.3.2: Please use the native JSON object instead of JSON 3
warning protractor > webdriver-manager > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning sass-lint > eslint > file-entry-cache > flat-cache > circular-json@0.3.3: CircularJSON is in maintenance only, flatted is its successor.
[2/4] Fetching packages...
warning mini-css-extract-plugin@0.4.5: Invalid bin field for "mini-css-extract-plugin".
warning url-loader@1.1.2: Invalid bin field for "url-loader".
info fsevents@1.2.13: The platform "linux" is incompatible with this module.
info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.3.2: The platform "linux" is incompatible with this module.
info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
Done in 24.00s.
```

File: build.gradle [hunk 1]
```diff
@@ -189,37 +189,32 @@ dockerCompose {
 
 task installNodeModules(type: Exec) {
     workingDir = uiWorkingDir
-    commandLine  'npm', 'install'
+    commandLine  'yarn', 'install'
```
Comment (rkennel) [inline build.gradle:192]: Good catch, forgot to make that change during the switch to yarn

File: ui/package.json [hunk 2]
```diff
@@ -1,35 +1,35 @@
...
   "scripts": {
     "ng": "ng",
     "start": "ng serve --configuration local",
     "build": "ng build",
     "build-prod": "ng build --prod --build-optimizer --output-path=../api/src/main/resources/static",
+    "test": "yarn unit-watch",
     "unit": "jest",
     "unit-watch": "jest --watchAll",
     "unit-coverage": "jest --coverage=true",
     "lint": "ng lint",
```
Comment (rkennel) [inline ui/package.json:13]: I recommend using the --cache setting to speed up linting

https://duncanleung.com/why-slow-plugin-typescript-eslint-performance-issues/

Comment (rkennel) [reply ui/package.json:13]: On my 12 core processor linux box linting takes 30 seconds.  with --cache setting and no changes it only takes 1 second

**Comment by rkennel** [comment]
Looks great, I like the inclusion of prettier and the pre-commit hook.  Take a look at my comment on the cache setting and if you agree please make the change.

**Review by rkennel — CHANGES_REQUESTED** [review]
Add --cache setting to lint to dramatically improve linting times

**Comment by adandach** [comment]
Caching has been added to the linter.

**Review by rkennel — APPROVED** [review]

**Pull request merged**